File: projects/rpn2tex/migrations/java-module-by-module-1/src/test/java/com/rpn2tex/IOContractTest.java
package com.rpn2tex;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Integration tests validating the I/O contract from the migration specification.
 *
 * <p>These tests verify end-to-end behavior: Lexer → Parser → LaTeXGenerator.
 * All test cases are taken from the migration spec Phase 0 I/O contract.
 */
@DisplayName("I/O Contract Integration Tests")
class IOContractTest {

    /**
     * Helper method to process RPN input through the full pipeline.
     *
     * @param input RPN expression string
     * @return LaTeX output string
     * @throws RpnException if lexing or parsing fails
     */
    private String processRPN(String input) throws RpnException {
        Lexer lexer = new Lexer(input);
        List<Token> tokens = lexer.tokenize();
        Parser parser = new Parser(tokens);
        Expr ast = parser.parse();
        LaTeXGenerator generator = new LaTeXGenerator();
        return generator.generate(ast);
    }

    @ParameterizedTest
    @CsvSource(delimiter = '|', value = {
        "5 3 +              | $5 + 3$",
        "5 3 -              | $5 - 3$",
        "4 7 *              | $4 \\times 7$",
        "10 2 /             | $10 \\div 2$",
        "5 3 + 2 *          | $( 5 + 3 ) \\times 2$",
        "5 3 * 2 +          | $5 \\times 3 + 2$",
        "10 2 / 5 *         | $10 \\div 2 \\times 5$",
        "5 3 - 2 -          | $5 - 3 - 2$",
        "100 10 / 5 / 2 /   | $100 \\div 10 \\div 5 \\div 2$",
        "1 2 + 3 + 4 +      | $1 + 2 + 3 + 4$",
        "2 3 4 * +          | $2 + 3 \\times 4$",
        "2 3 + 4 *          | $( 2 + 3 ) \\times 4$",
        "2 3 4 + *          | $2 \\times ( 3 + 4 )$",
        "2 3 * 4 +          | $2 \\times 3 + 4$",
        "3.14 2 *           | $3.14 \\times 2$",
        "1.5 0.5 +          | $1.5 + 0.5$",
        "1 2 + 3 4 + *      | $( 1 + 2 ) \\times ( 3 + 4 )$",
        "10 2 / 3 + 4 *     | $( 10 \\div 2 + 3 ) \\times 4$"
    })
    @DisplayName("should generate correct LaTeX for valid RPN expressions")
    void testValidExpressions(String input, String expectedOutput) throws RpnException {
        String result = processRPN(input);
        assertEquals(expectedOutput, result,
            String.format("Input: '%s' should produce: '%s'", input, expectedOutput));
    }

    @Test
    @DisplayName("should generate LaTeX for basic addition: 5 3 +")
    void testBasicAddition() throws RpnException {
        String result = processRPN("5 3 +");
        assertEquals("$5 + 3$", result);
    }

    @Test
    @DisplayName("should generate LaTeX for basic subtraction: 5 3 -")
    void testBasicSubtraction() throws RpnException {
        String result = processRPN("5 3 -");
        assertEquals("$5 - 3$", result);
    }

    @Test
    @DisplayName("should generate LaTeX for basic multiplication: 4 7 *")
    void testBasicMultiplication() throws RpnException {
        String result = processRPN("4 7 *");
        assertEquals("$4 \\times 7$", result);
    }

    @Test
    @DisplayName("should generate LaTeX for basic division: 10 2 /")
    void testBasicDivision() throws RpnException {
        String result = processRPN("10 2 /");
        assertEquals("$10 \\div 2$", result);
    }

    @Test
    @DisplayName("should throw exception for unsupported exponentiation: 2 3 ^")
    void testExponentiationNotSupported() {
        RpnException exception = assertThrows(RpnException.class, () -> {
            processRPN("2 3 ^");
        });
        assertTrue(exception.getMessage().contains("Unexpected character '^'"),
            "Error message should mention unexpected character '^'");
    }

    @Test
    @DisplayName("should add parentheses for precedence: 5 3 + 2 *")
    void testPrecedenceWithParentheses() throws RpnException {
        String result = processRPN("5 3 + 2 *");
        assertEquals("$( 5 + 3 ) \\times 2$", result);
    }

    @Test
    @DisplayName("should handle operator precedence without parens: 5 3 * 2 +")
    void testPrecedenceNoParentheses() throws RpnException {
        String result = processRPN("5 3 * 2 +");
        assertEquals("$5 \\times 3 + 2$", result);
    }

    @Test
    @DisplayName("should handle left-associative operations: 10 2 / 5 *")
    void testLeftAssociativeOps() throws RpnException {
        String result = processRPN("10 2 / 5 *");
        assertEquals("$10 \\div 2 \\times 5$", result);
    }

    @Test
    @DisplayName("should handle multiple subtractions: 5 3 - 2 -")
    void testMultipleSubtractions() throws RpnException {
        String result = processRPN("5 3 - 2 -");
        assertEquals("$5 - 3 - 2$", result);
    }

    @Test
    @DisplayName("should handle multiple divisions: 100 10 / 5 / 2 /")
    void testMultipleDivisions() throws RpnException {
        String result = processRPN("100 10 / 5 / 2 /");
        assertEquals("$100 \\div 10 \\div 5 \\div 2$", result);
    }

    @Test
    @DisplayName("should handle multiple additions: 1 2 + 3 + 4 +")
    void testMultipleAdditions() throws RpnException {
        String result = processRPN("1 2 + 3 + 4 +");
        assertEquals("$1 + 2 + 3 + 4$", result);
    }

    @Test
    @DisplayName("should handle multiplication with higher precedence: 2 3 4 * +")
    void testMultiplicationHigherPrecedence() throws RpnException {
        String result = processRPN("2 3 4 * +");
        assertEquals("$2 + 3 \\times 4$", result);
    }

    @Test
    @DisplayName("should add parens for addition before mult: 2 3 + 4 *")
    void testAdditionBeforeMultiplication() throws RpnException {
        String result = processRPN("2 3 + 4 *");
        assertEquals("$( 2 + 3 ) \\times 4$", result);
    }

    @Test
    @DisplayName("should add parens for addition as second operand: 2 3 4 + *")
    void testAdditionAsSecondOperand() throws RpnException {
        String result = processRPN("2 3 4 + *");
        assertEquals("$2 \\times ( 3 + 4 )$", result);
    }

    @Test
    @DisplayName("should not add unnecessary parens: 2 3 * 4 +")
    void testNoUnnecessaryParens() throws RpnException {
        String result = processRPN("2 3 * 4 +");
        assertEquals("$2 \\times 3 + 4$", result);
    }

    @Test
    @DisplayName("should handle floating-point multiplication: 3.14 2 *")
    void testFloatingPointMultiplication() throws RpnException {
        String result = processRPN("3.14 2 *");
        assertEquals("$3.14 \\times 2$", result);
    }

    @Test
    @DisplayName("should handle floating-point addition: 1.5 0.5 +")
    void testFloatingPointAddition() throws RpnException {
        String result = processRPN("1.5 0.5 +");
        assertEquals("$1.5 + 0.5$", result);
    }

    @Test
    @DisplayName("should handle both operands parenthesized: 1 2 + 3 4 + *")
    void testBothOperandsParenthesized() throws RpnException {
        String result = processRPN("1 2 + 3 4 + *");
        assertEquals("$( 1 + 2 ) \\times ( 3 + 4 )$", result);
    }

    @Test
    @DisplayName("should handle complex mixed expression: 10 2 / 3 + 4 *")
    void testComplexMixedExpression() throws RpnException {
        String result = processRPN("10 2 / 3 + 4 *");
        assertEquals("$( 10 \\div 2 + 3 ) \\times 4$", result);
    }

    @Test
    @DisplayName("should reject exponentiation in complex expression: 2 3 ^ 4 *")
    void testExponentiationInComplexExpression() {
        RpnException exception = assertThrows(RpnException.class, () -> {
            processRPN("2 3 ^ 4 *");
        });
        assertTrue(exception.getMessage().contains("Unexpected character '^'"));
    }

    @Test
    @DisplayName("should reject multiple exponentiations: 2 3 4 ^ ^")
    void testMultipleExponentiations() {
        RpnException exception = assertThrows(RpnException.class, () -> {
            processRPN("2 3 4 ^ ^");
        });
        assertTrue(exception.getMessage().contains("Unexpected character '^'"));
    }

    @Test
    @DisplayName("should handle negative numbers")
    void testNegativeNumbers() throws RpnException {
        String result = processRPN("-5 3 +");
        assertEquals("$-5 + 3$", result);
    }

    @Test
    @DisplayName("should handle negative floating-point numbers")
    void testNegativeFloatingPoint() throws RpnException {
        String result = processRPN("-3.14 2 *");
        assertEquals("$-3.14 \\times 2$", result);
    }

    @Test
    @DisplayName("should preserve number format from input")
    void testNumberFormatPreservation() throws RpnException {
        // Leading zero
        String result1 = processRPN("0.5 0.5 +");
        assertEquals("$0.5 + 0.5$", result1);

        // Large numbers
        String result2 = processRPN("1000 2000 +");
        assertEquals("$1000 + 2000$", result2);
    }

    @Test
    @DisplayName("should handle whitespace variations")
    void testWhitespaceVariations() throws RpnException {
        // Multiple spaces
        String result1 = processRPN("5   3   +");
        assertEquals("$5 + 3$", result1);

        // Tabs
        String result2 = processRPN("5\t3\t+");
        assertEquals("$5 + 3$", result2);

        // Mixed whitespace
        String result3 = processRPN("5 \t 3  \t+");
        assertEquals("$5 + 3$", result3);
    }

    @Test
    @DisplayName("should handle expressions with newlines")
    void testNewlineHandling() throws RpnException {
        String result = processRPN("5 3 +\n2 *");
        assertEquals("$( 5 + 3 ) \\times 2$", result);
    }

    @Test
    @DisplayName("LaTeX output should always be wrapped in dollar signs")
    void testMathModeWrapping() throws RpnException {
        String result = processRPN("42");
        assertTrue(result.startsWith("$"));
        assertTrue(result.endsWith("$"));
    }

    @Test
    @DisplayName("should properly escape LaTeX commands")
    void testLatexCommandEscaping() throws RpnException {
        String multResult = processRPN("2 3 *");
        assertTrue(multResult.contains("\\times"), "Multiplication should use \\times");

        String divResult = processRPN("10 2 /");
        assertTrue(divResult.contains("\\div"), "Division should use \\div");
    }
}

File: projects/rpn2tex/migrations/rust-feature-by-feature-3/tests/io_contract_test.rs
//! I/O Contract Tests - Verifies exact behavior from Phase 1 specification.
//!
//! These tests ensure that the implementation matches the expected behavior
//! documented in the feature specification.

use rpn2tex::compile;

/// Test Case: Numbers - Single integer
/// Input: "5"
/// Expected: "$5$"
#[test]
fn io_contract_numbers_single_integer() {
    let input = "5";
    let expected = "$5$";
    let result = compile(input);
    assert!(result.is_ok(), "Expected Ok, got Err: {:?}", result);
    assert_eq!(
        result.unwrap(),
        expected,
        "Output does not match expected for input: {input}"
    );
}

/// Test Case: Numbers - Decimal number
/// Input: "3.14"
/// Expected: "$3.14$"
#[test]
fn io_contract_numbers_decimal() {
    let input = "3.14";
    let expected = "$3.14$";
    let result = compile(input);
    assert!(result.is_ok(), "Expected Ok, got Err: {:?}", result);
    assert_eq!(
        result.unwrap(),
        expected,
        "Output does not match expected for input: {input}"
    );
}

/// Error Case: Empty expression
/// Input: (empty)
/// Expected Error: "Error: Empty expression"
#[test]
fn io_contract_error_empty_expression() {
    let input = "";
    let result = compile(input);
    assert!(result.is_err(), "Expected Err, got Ok for empty input");
    let error = result.unwrap_err();
    assert_eq!(error, "Error: Empty expression");
}

/// Error Case: Unexpected character
/// Input: "5 @ 3"
/// Expected Error: Contains "Unexpected character '@'"
#[test]
fn io_contract_error_unexpected_character() {
    let input = "5 @";
    let result = compile(input);
    assert!(
        result.is_err(),
        "Expected Err, got Ok for input with invalid character"
    );
    let error = result.unwrap_err();
    assert!(
        error.contains("Unexpected character"),
        "Error message should mention unexpected character"
    );
    assert!(error.contains("@"), "Error message should include '@'");
}

/// Test Case: Addition - Basic addition
/// Input: "5 3 +"
/// Expected: "$5 + 3$"
#[test]
fn io_contract_addition_basic() {
    let input = "5 3 +";
    let expected = "$5 + 3$";
    let result = compile(input);
    assert!(result.is_ok(), "Expected Ok, got Err: {:?}", result);
    assert_eq!(
        result.unwrap(),
        expected,
        "Output does not match expected for input: {input}"
    );
}

/// Test Case: Addition - Chained additions (left-associative)
/// Input: "1 2 + 3 + 4 +"
/// Expected: "$1 + 2 + 3 + 4$"
#[test]
fn io_contract_addition_chained() {
    let input = "1 2 + 3 + 4 +";
    let expected = "$1 + 2 + 3 + 4$";
    let result = compile(input);
    assert!(result.is_ok(), "Expected Ok, got Err: {:?}", result);
    assert_eq!(
        result.unwrap(),
        expected,
        "Output does not match expected for input: {input}"
    );
}

/// Error Case: Addition with insufficient operands
/// Input: "5 +"
/// Expected Error: Contains "requires two operands"
#[test]
fn io_contract_error_addition_insufficient_operands() {
    let input = "5 +";
    let result = compile(input);
    assert!(
        result.is_err(),
        "Expected Err, got Ok for addition with insufficient operands"
    );
    let error = result.unwrap_err();
    assert!(
        error.contains("requires two operands"),
        "Error message should mention 'requires two operands', got: {error}"
    );
}

/// Error Case: Addition with no operands
/// Input: "+"
/// Expected Error: Contains "requires two operands"
#[test]
fn io_contract_error_addition_no_operands() {
    let input = "+";
    let result = compile(input);
    assert!(
        result.is_err(),
        "Expected Err, got Ok for addition with no operands"
    );
    let error = result.unwrap_err();
    assert!(
        error.contains("requires two operands"),
        "Error message should mention 'requires two operands', got: {error}"
    );
}

/// Test Case: Subtraction - Basic subtraction
/// Input: "5 3 -"
/// Expected: "$5 - 3$"
#[test]
fn io_contract_subtraction_basic() {
    let input = "5 3 -";
    let expected = "$5 - 3$";
    let result = compile(input);
    assert!(result.is_ok(), "Expected Ok, got Err: {:?}", result);
    assert_eq!(
        result.unwrap(),
        expected,
        "Output does not match expected for input: {input}"
    );
}

/// Test Case: Subtraction - Chained subtractions (left-associative)
/// Input: "5 3 - 2 -"
/// Expected: "$5 - 3 - 2$"
#[test]
fn io_contract_subtraction_chained() {
    let input = "5 3 - 2 -";
    let expected = "$5 - 3 - 2$";
    let result = compile(input);
    assert!(result.is_ok(), "Expected Ok, got Err: {:?}", result);
    assert_eq!(
        result.unwrap(),
        expected,
        "Output does not match expected for input: {input}"
    );
}

/// Test Case: Multiplication - Basic multiplication
/// Input: "4 7 *"
/// Expected: "$4 \times 7$"
#[test]
fn io_contract_multiplication_basic() {
    let input = "4 7 *";
    let expected = r"$4 \times 7$";
    let result = compile(input);
    assert!(result.is_ok(), "Expected Ok, got Err: {:?}", result);
    assert_eq!(
        result.unwrap(),
        expected,
        "Output does not match expected for input: {input}"
    );
}

/// Test Case: Multiplication - Multiplication with addition
/// Input: "2 3 4 * +"
/// Expected: "$2 + 3 \times 4$"
/// Note: No precedence handling yet, so no parentheses around 3*4
#[test]
fn io_contract_multiplication_with_addition() {
    let input = "2 3 4 * +";
    let expected = r"$2 + 3 \times 4$";
    let result = compile(input);
    assert!(result.is_ok(), "Expected Ok, got Err: {:?}", result);
    assert_eq!(
        result.unwrap(),
        expected,
        "Output does not match expected for input: {input}"
    );
}

/// Error Case: Multiplication with insufficient operands
/// Input: "5 *"
/// Expected Error: Contains "requires two operands"
#[test]
fn io_contract_error_multiplication_insufficient_operands() {
    let input = "5 *";
    let result = compile(input);
    assert!(
        result.is_err(),
        "Expected Err, got Ok for multiplication with insufficient operands"
    );
    let error = result.unwrap_err();
    assert!(
        error.contains("requires two operands"),
        "Error message should mention 'requires two operands', got: {error}"
    );
}

/// Error Case: Multiplication with no operands
/// Input: "*"
/// Expected Error: Contains "requires two operands"
#[test]
fn io_contract_error_multiplication_no_operands() {
    let input = "*";
    let result = compile(input);
    assert!(
        result.is_err(),
        "Expected Err, got Ok for multiplication with no operands"
    );
    let error = result.unwrap_err();
    assert!(
        error.contains("requires two operands"),
        "Error message should mention 'requires two operands', got: {error}"
    );
}
